File: 18-4sum/4sum.cpp
#define ll long long
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> quads;
        for(int i=0; i<nums.size(); i++){
            if (i > 0 && nums[i] == nums[i-1]) continue;
            int a = nums[i];
            for(int j=i+1; j<nums.size(); j++){
                if (j > i+1 && nums[j] == nums[j-1]) continue;
                int b = nums[j];
                int l = j+1, r = nums.size()-1;
                while (l < r){
                    int c = nums[l], d = nums[r];
                    ll sum = (ll)a + (ll)b + (ll)c + (ll)d;
                    
                    if (sum == target){
                        quads.push_back({a, b, c, d});
                        l++;
                        r--;

                        while(l < r && nums[l] == nums[l-1]) l++;
                        while(l < r && nums[r] == nums[r+1]) r--;
                    } else if (sum > target){
                        r--;
                    } else {
                        l++;
                    }
                }
            }
        }
        return quads;
    }
};

File: 18-4sum/4sum.go
func fourSum(nums []int, target int) [][]int {
    sort.Ints(nums)
    var ans [][]int
    for i := 0; i < len(nums); i++ {
        if i > 0 && nums[i] == nums[i-1] {
            continue
        }
        first := nums[i]
        for j := i + 1; j < len(nums); j++ {
            if j > i+1 && nums[j] == nums[j-1] {
                continue
            }
            second := nums[j]
            l, r := j + 1, len(nums)-1
            for l < r {
                sum := first + second + nums[l] + nums[r]
                if sum == target {
                    temp := []int{first, second, nums[l], nums[r]}
                    ans = append(ans, temp)


                    l++
                    r--

                    for l < r && nums[l] == nums[l-1] {
                        l++
                    }
                    for l < r && nums[r] == nums[r+1] {
                        r--
                    }
                } else if sum > target {
                    r--
                } else {
                    l++
                }
            }
        }
    }

    return ans
}